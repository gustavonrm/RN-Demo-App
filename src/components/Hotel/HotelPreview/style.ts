import { StyleSheet } from 'react-native';
import { BLACK, GREY, PINK, WHITE } from '../../../constants/colors';

const style = StyleSheet.create({
  body: {
    marginHorizontal: 15,
    marginTop: 10,
    marginBottom: 100,
  },
  image: {
    flex: 1,
    width: 390,
    height: 250,
    borderRadius: 10,
    margin: 1,
  },
  stars: {
    flexDirection: 'row',
    marginTop: 10,
    alignItems: 'center',
  },
  title: {
    fontWeight: 'bold',
    color: BLACK,
    fontSize: 24,
  },
  subtitle: {
    fontWeight: '600',
    color: BLACK,
    fontSize: 20,
    marginTop: 2,
    marginBottom: 5,
  },
  text: {
    fontSize: 16,
    color: BLACK,
    fontWeight: '400',
    alignItems: 'center',
    marginVertical: 3,
  },
  divider: {
    borderWidth: 1,
    borderColor: GREY,
    marginVertical: 7,
  },
  container: {
    backgroundColor: WHITE,
  },
  actionButtonsContainer: {
    flexDirection: 'row',
    flex: 1,
    marginTop: 5,
    marginHorizontal: -10,
  },
  bottomContainer: {
    position: 'absolute',
    flex: 1,
    flexDirection: 'row',
    bottom: 0,
    backgroundColor: PINK,
    flexGrow: 1,
    width: '100%',
    padding: 10,
    paddingBottom: 20,
    paddingHorizontal: 15,
    alignItems: 'flex-start',
    justifyContent: 'space-between',
  },
  bookButton: {
    minWidth: 150,
  },
});

export default style;
