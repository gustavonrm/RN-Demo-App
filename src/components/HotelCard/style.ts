import { StyleSheet } from 'react-native';
import { GREY, PINK, WHITE } from '../../constants/colors';

const style = StyleSheet.create({
    container: {
        flex: 1,
        borderWidth:1,
        borderColor: GREY,
        borderRadius: 10,
        backgroundColor: WHITE,
        marginVertical: 10,
    },
    image:{
        flex:1,
        width: undefined,
        height: undefined,
        borderRadius: 10,
    },
    priceContainer:{
     position: 'absolute',
       fontWeight: 'bold',
       right: 0,
       bottom: 0,
       margin: 10,
       alignItems: 'flex-end',
    },
    priceText:{
          fontWeight: 'bold',
          color: WHITE,
            fontSize: 24,
          textShadowColor: PINK,
          textShadowOffset: {width: -1, height: 1},
          textShadowRadius: 4,
       },
    priceText2:{
        fontWeight: '100!important',
        color: WHITE,
        fontSize: 14,
        textShadowColor: PINK,
        textShadowOffset: {width: -1, height: 1},
        textShadowRadius: 4,
     },
     priceText3:{
        fontWeight: 'bold',
        color: WHITE,
          fontSize: 14,
        textShadowColor: PINK,
        textShadowOffset: {width: -1, height: 1},
        textShadowRadius: 4,
        marginTop: 5,
     },
    top:{
        borderRadius: 10,
        height: 200,
        flex: 1,
        flexDirection: 'column',
    },
    bottom:{
        flex: 1,
        flexDirection: 'column',
        margin: 10,
    },
});

export default style;
