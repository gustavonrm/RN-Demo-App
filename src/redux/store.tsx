import { configureStore, combineReducers } from '@reduxjs/toolkit';
import hotelsSlice from './slices/hotels.slice';
import { hotelsApi } from './apis/hotels.api';
import filtersSlice from './slices/filters.slice';

const rootReducer = combineReducers({
  hotels: hotelsSlice,
  filters: filtersSlice,
  [hotelsApi.reducerPath]: hotelsApi.reducer,
});

export const setupStore = (preloadedState?: Partial<RootState>) =>
  configureStore({
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(hotelsApi.middleware),
    preloadedState,
  });

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
