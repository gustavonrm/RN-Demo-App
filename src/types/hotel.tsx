export interface Hotel {
    id: number,
    name: string,
    location: Location,
    stars: number,
    checkIn: FromToHours,
    checkOut: FromToHours,
    contact: Contact,
    gallery: string[],
    userRating: number,
    price: string,
    currency: string,
}

export interface Location {
    address: string,
    city: string,
    latitude: number,
    longitude: number
}

export interface FromToHours {
    from: string,
    to: string,
}

export interface Contact {
    phoneNumber: string,
    email: string,
}
