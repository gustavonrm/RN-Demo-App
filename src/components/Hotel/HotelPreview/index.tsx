import React, { useCallback, useState } from 'react';
import { ScrollView, Text, View, Image, Linking, Alert } from 'react-native';
import { selectHotel } from '../../../redux/slices/hotels.slice';
import withError from '../../HOC/withError';
import { useSelector } from 'react-redux';
import style from './style';
import { FontAwesomeIcon } from '@fortawesome/react-native-fontawesome';
import { PINK, WHITE, YELLOW } from '../../../constants/colors';
import ErrorView from '../../common/ErrorView';
import ActionButton from '../../common/ActionButton';
import TextIcon from '../../common/TextIcon';

type HotelPreviewProps = {
  id: number;
};

const HotelPreview = ({ id }: HotelPreviewProps) => {
  const hotel = useSelector((state) => selectHotel(id)(state));
  const [booked, setBooked] = useState(false);

  if (!hotel) return <ErrorView message={'Error retriving the hotel'} />;

  const startCall = useCallback(
    (phone: string) => () => {
      Linking.openURL(`tel:${phone}`).catch((error) => {
        Alert.alert('Error staring a call!', error);
      });
    },
    []
  );

  const sendEmail = useCallback(
    (email: string) => () => {
      Linking.openURL(`mailto:${email}`).catch((error) => {
        Alert.alert('Error sending an Email!', error);
      });
    },
    []
  );

  return (
    <>
      <ScrollView
        testID="hotelPreviewTestId"
        style={style.container}
        showsVerticalScrollIndicator={false}
      >
        <ScrollView horizontal persistentScrollbar>
          {hotel?.gallery.map((image, index) => (
            <Image
              key={index}
              style={style.image}
              source={{
                uri:
                  image ||
                  'https://upload.wikimedia.org/wikipedia/commons/1/14/No_Image_Available.jpg?20200913095930',
              }}
            />
          ))}
        </ScrollView>
        <View style={style.body}>
          <Text style={style.title}>{hotel?.name}</Text>
          <View style={style.stars}>
            {Array.from({ length: hotel.stars }).map((_, index) => (
              <FontAwesomeIcon testID={'starTestId'} key={index} icon="star" color={YELLOW} />
            ))}
            <Text style={style.text}>{` | ${hotel.userRating} user rating`}</Text>
          </View>

          <View style={style.divider} />

          <Text style={style.subtitle}>{'Address Information:'}</Text>
          <TextIcon
            icon={'location-dot'}
            text={`${hotel.location.city}, ${hotel.location.address}`}
          />
          <Text style={style.text}>{`Latitude: ${hotel.location.latitude}`}</Text>
          <Text style={style.text}>{`Longitude: ${hotel.location.latitude}`}</Text>
          <TextIcon icon={'clock'} text={`Check in: ${hotel.checkIn.from}-${hotel.checkIn.to}`} />
          <TextIcon
            icon={'clock'}
            text={`Check out: ${hotel.checkOut.from}-${hotel.checkOut.to}`}
          />

          <View style={style.divider} />

          <Text style={style.subtitle}>{'Contact Information:'}</Text>
          <TextIcon icon={'phone'} text={`Phone Number: ${hotel.contact.phoneNumber}`} />
          <TextIcon icon={'envelope'} text={`Email: ${hotel.contact.email}`} />

          <View style={style.actionButtonsContainer}>
            <ActionButton secondary onPress={startCall(hotel.contact.phoneNumber)}>
              <FontAwesomeIcon icon={'phone'} />
            </ActionButton>
            <ActionButton secondary onPress={sendEmail(hotel.contact.email)}>
              <FontAwesomeIcon icon={'envelope'} />
            </ActionButton>
          </View>
          <View>
            <View style={style.divider} />
          </View>
        </View>
      </ScrollView>
      <View style={style.bottomContainer}>
        <View>
          <Text style={{ ...style.subtitle, color: WHITE }}>{`€ ${hotel.price} total`}</Text>
          <Text
            style={{ ...style.text, color: WHITE, fontSize: 12 }}
          >{`€ ${hotel.price} /person`}</Text>
        </View>
        <View style={style.bookButton}>
          <ActionButton secondary onPress={() => setBooked(!booked)}>
            {booked ? (
              <Text style={{ ...style.text, color: PINK }}>{`Cancel Booking`}</Text>
            ) : (
              <Text style={{ ...style.text, color: PINK }}>{`Book Hotel`}</Text>
            )}
          </ActionButton>
        </View>
      </View>
    </>
  );
};

export default withError(HotelPreview, (state, props) => selectHotel(props.id)(state));
