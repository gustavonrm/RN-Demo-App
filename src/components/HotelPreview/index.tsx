import React from 'react';
import { ScrollView, Text, View, Image } from 'react-native';
import { selectHotel } from '../../redux/slices/hotels.slice';
import withError from '../HOC/withError';
import { useSelector } from 'react-redux';
import style from './style';

type HotelPreviewProps = {
  id: number;
};

const HotelPreview = ({ id }: HotelPreviewProps) => {
  const hotel = useSelector((state) => selectHotel(id)(state));

  console.log(hotel?.gallery[0]);
  return (
    <View>
      <ScrollView horizontal>
        {hotel?.gallery.map((image) => (
          <Image
            style={style.image}
            source={{
              uri:
                image ||
                'https://upload.wikimedia.org/wikipedia/commons/1/14/No_Image_Available.jpg?20200913095930',
            }}
          />
        ))}
      </ScrollView>
      <Text>{hotel?.name}</Text>
    </View>
  );
};

export default withError(HotelPreview, (state, props) => selectHotel(props.id)(state));
