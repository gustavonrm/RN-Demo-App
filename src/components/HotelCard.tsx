import React from 'react';
import { Image, Text, TouchableOpacity} from 'react-native';
import { Hotel } from '../types/hotel';
import { useNavigation } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';
import { RootStackParamList } from '../navigationTypes';

type HotelCardProps = {
    hotel: Hotel
}

const HotelCard = ({hotel}: HotelCardProps) => {

const navigation = useNavigation<StackNavigationProp<RootStackParamList>>();
  return (
  <TouchableOpacity onPress={() => navigation.navigate('Preview', {id: hotel.id})}>
    <Image style={{width: 100,height: 100}} source={{uri: hotel.gallery[0] || 'https://upload.wikimedia.org/wikipedia/commons/1/14/No_Image_Available.jpg?20200913095930'}}/>
    <Text>{hotel.name}</Text>
    <Text>{hotel.location.city} - {hotel.location.address}</Text>
    <Text>{hotel.price} {hotel.currency}</Text>
    </TouchableOpacity>
  );
};

export default HotelCard;
