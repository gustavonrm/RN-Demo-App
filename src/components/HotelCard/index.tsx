import React from 'react';
import { Image, Text, TouchableOpacity, View} from 'react-native';
import { Hotel } from '../../types/hotel';
import { useNavigation } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';
import { RootStackParamList } from '../navigationTypes';
import { FontAwesomeIcon } from '@fortawesome/react-native-fontawesome';
import style from './style';

type HotelCardProps = {
    hotel: Hotel
}

const simpleCurrencyFormatter = (price: string, currency: string) => {
  let currencySymbol = currency === 'EUR' ? '€' : '$';
  return currencySymbol + price;
};

const HotelCard = ({hotel}: HotelCardProps) => {

  const navigation = useNavigation<StackNavigationProp<RootStackParamList>>();

  return (
  <TouchableOpacity style={style.container} onPress={() => navigation.navigate('Preview', {id: hotel.id})}>
    <View style={style.top}>
      <Image style={style.image} source={{uri: hotel.gallery[0] || 'https://upload.wikimedia.org/wikipedia/commons/1/14/No_Image_Available.jpg?20200913095930'}}/>
      <View style={style.priceContainer}>
        <Text style={style.priceText}>
          {`${hotel.currency === 'EUR' ? '€' : '$'}${hotel.price}`}<Text style={style.priceText2}>{' /person'}</Text>
        </Text>
        <Text style={style.priceText3}>
        {`${hotel.currency === 'EUR' ? '€' : '$'}${hotel.price}`} {'total'}
        </Text>
      </View>
    </View>
    <View style={style.bottom}>
      <Text>{hotel.location.city} - {hotel.location.address}</Text>
      <Text>{hotel.name}</Text>
      <FontAwesomeIcon icon="star" />
      <Text>{hotel.stars}</Text><Text>{hotel.userRating}</Text>
      <Text>{hotel.checkIn.from}</Text><Text>{hotel.checkIn.to}</Text><Text>{hotel.checkOut.from}</Text><Text>{hotel.checkOut.to}</Text>
    </View>


    </TouchableOpacity>
  );
};

export default HotelCard;
