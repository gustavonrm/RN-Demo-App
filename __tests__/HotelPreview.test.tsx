import React from 'react';
import 'react-native';
import { render, screen, userEvent } from '@testing-library/react-native';
import HotelCard from '../src/components/Hotel/HotelCard';
import { hotel } from './mocks/hotel';
import HotelPreview from '../src/components/Hotel/HotelPreview';
import { renderWithProviders } from './utils/test-utils';
import { hotels } from './mocks/hotels';
import { setupStore } from '../src/redux/store';
import { setHotels } from '../src/redux/slices/hotels.slice';

describe('HotelCard Element', () => {
  it('renders hotel card correctly', () => {
    const store = setupStore();
    store.dispatch(setHotels(hotels));

    renderWithProviders(<HotelPreview id={hotels[0].id} />, { store });
    expect(screen.getByTestId('hotelPreviewTestId')).toBeTruthy();

    expect(screen.getByText(hotel.name)).toBeTruthy();
    expect(screen.getAllByTestId('starTestId').length).toBe(hotel.stars);
    expect(screen.getByText(`| ${hotel.userRating} user rating`)).toBeTruthy();

    expect(screen.getByText('Address Information:')).toBeTruthy();
    expect(screen.getByText(`${hotel.location.city}, ${hotel.location.address}`)).toBeTruthy();
    expect(screen.getByText(`Latitude: ${hotel.location.latitude}`)).toBeTruthy();
    expect(screen.getByText(`Longitude: ${hotel.location.latitude}`)).toBeTruthy();

    expect(screen.getByText(`Check in: ${hotel.checkIn.from}-${hotel.checkIn.to}`)).toBeTruthy();
    expect(screen.getByText(`Check out: ${hotel.checkOut.from}-${hotel.checkOut.to}`)).toBeTruthy();

    expect(screen.getByText('Contact Information:')).toBeTruthy();
    expect(screen.getByText(`Phone Number: ${hotel.contact.phoneNumber}`)).toBeTruthy();
    expect(screen.getByText(`Email: ${hotel.contact.email}`)).toBeTruthy();

    expect(screen.getByText(`Book Hotel`)).toBeTruthy();
    expect(screen.getByText(`€${hotel.price} /person`)).toBeTruthy();
    expect(screen.getByText(`€${hotel.price} total`)).toBeTruthy();
  });

  it('book hotel', async () => {
    const user = userEvent.setup();
    const store = setupStore();
    store.dispatch(setHotels(hotels));

    renderWithProviders(<HotelPreview id={hotels[0].id} />, { store });
    expect(screen.getByTestId('hotelPreviewTestId')).toBeTruthy();

    expect(screen.getByText(`Book Hotel`)).toBeTruthy();
    await user.press(screen.getByText(`Book Hotel`));
    expect(screen.getByText(`Cancel Booking`)).toBeTruthy();
  });

  it('renders error if no data', () => {
    const store = setupStore();
    store.dispatch(setHotels([]));

    renderWithProviders(<HotelPreview id={999} />, { store });
    expect(screen.getByTestId('errorViewTestId')).toBeTruthy();
  });
});
