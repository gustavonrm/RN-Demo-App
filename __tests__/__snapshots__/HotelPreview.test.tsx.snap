// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HotelCard Element snapshot test 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
  testID="errorViewTestId"
>
  <Text
    style={
      {
        "fontSize": 24,
        "fontWeight": "bold",
      }
    }
  >
    Error Loading Data
  </Text>
</View>
`;
